instance-scheduler/environments.go

```diff
@@ -0,0 +1,154 @@
+    }
+    defer resp.Body.Close()
+
+    if resp.StatusCode != http.StatusOK {
+        return nil, fmt.Errorf("non-200 status code: %d", resp.StatusCode)
+    }
+
+    body, err := io.ReadAll(resp.Body)
+    if err != nil {
+        return nil, fmt.Errorf("failed to read response body: %w", err)
+    }
+
+    var content JSONFileContent
+    err = json.Unmarshal(body, &content)
+	  if err != nil {
+		  return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	  }
+
+    return content, nil
+}
+
+// fetches the environments JSON data from GitHub
+func fetchGitHubData(baseURL, repoOwner, repoName, branch, directory string) ([]byte, error) {
+
+    //u, err := url.Parse(path.Join(baseURL, repoOwner, repoName, "contents", directory))
+    u, err := url.Parse(fmt.Sprintf("%s/%s/%s/contents/%s", baseURL, repoOwner, repoName, directory))
+    if err != nil {
+        return nil, fmt.Errorf("failed to parse URL: %w", err)
+    }
+
+    query := u.Query()
+    query.Set("ref", branch)
+    u.RawQuery = query.Encode()
+
+    // Print the constructed URL for debugging
+    fmt.Println("Constructed URL:", u.String())
+
+    // Create a new HTTP request
+    req, err := http.NewRequest("GET", u.String(), nil)
+    if err != nil {
+        return nil, fmt.Errorf("failed to create HTTP request: %w", err)
+    }
+
+    // Set the User-Agent header (GitHub API requires a User-Agent header)
+    req.Header.Set("User-Agent", "Go-http-client")
+
+    // Perform the HTTP request
+    client := &http.Client{}
+    resp, err := client.Do(req)
+    if err != nil {
+        return nil, fmt.Errorf("failed to fetch directory listing: %w", err)
+    }
+    defer resp.Body.Close()
+
+    if resp.StatusCode != http.StatusOK {
+        return nil, fmt.Errorf("non-200 status code: %d", resp.StatusCode)
+    }
+
+    body, err := io.ReadAll(resp.Body)
+    if err != nil {
+        return nil, fmt.Errorf("failed to read response body: %w", err)
+    }
+
+    return body, nil
+}
+
+// processGitHubData processes the JSON data and returns a slice of GitHubFile
+func processGitHubData(body []byte) ([]GitHubFile, error) {
+    var files []GitHubFile
+    err := json.Unmarshal(body, &files)
+    if err != nil {
+        return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+    }
+    return files, nil
+}
+
+
+// hasInstanceSchedulerSkip checks if the instance_scheduler_skip field exists and contains "true"
+func hasInstanceSchedulerSkip(content gjson.Result) bool {
+    skip := content.Get("instance_scheduler_skip")
+    if skip.Exists() {
+        for _, value := range skip.Array() {
+            if value.String() == "true" {
+                return true
+            }
+        }
+    }
+    return false
+}
+
+// extractNames finds all "name" elements in the "environments" array, excluding those with instance_scheduler_skip or production
+func extractNames(content JSONFileContent, envName string) []string {
```

Comment:
Thinking this needs a more verbose name as it contains a lot of business logic 🤔 

Something along the lines of 👇 

```suggestion
func extractNonProductionAndNonSkippedEnvironmentNames(content JSONFileContent, envName string) []string {
```

In a later refactor, we might want to separate some of the code out to uplift the business logic up 🆙 Something along the lines of 👇 

```go
environments := filterEnvironmentsConfiguredToSkip(filterProductionEnvironments(getEnvironments()))
```